### PR TITLE
feat: add basic support for nix flakes

### PR DESCRIPTION
**Pull request by isubasinghe**
This adds a reproducible build environment for Argo Workflows. Argo Workflows can now be run in a few simple steps documented in the README.md. 

Why are there some go files changed here?
This was the outcome of running gofmt on the sources, sources can be reformatted by running "nix fmt". 
Currently this only supports formatting golang and nix files but can be extended to anything supported via clang-format, prettier and more. See this for more info: https://numtide.github.io/treefmt/formatters/

The files.go stub was added, this is a bit more deterministic than copying at build time. This will also make caching easier.

### CODE REVIEW

**Comment by JPZ13** [comment]
I'm running into an error with `devenv up`
```
(devenv) JPs-MacBook-Pro:argo-workflows jpzivalich$ devenv up
mkdir: /.devenv: Read-only file system
/nix/store/3hbmx5ka5k6m9hijylj0skri9wfp6rag-devenv-up: line 22: /.devenv/state/devenv.pid: No such file or directory
11:23:55 system                | ui.1 started (pid=2625)
11:23:55 system                | argo-server.1 started (pid=2626)
11:23:55 system                | workflow-controller.1 started (pid=2627)
11:23:55 workflow-controller.1 | time="2023-05-02T11:23:55+02:00" level=info msg="index config" indexWorkflowSemaphoreKeys=true
11:23:55 workflow-controller.1 | time="2023-05-02T11:23:55+02:00" level=info msg="cron config" cronSyncPeriod=10s
11:23:55 workflow-controller.1 | time="2023-05-02T11:23:55+02:00" level=info msg="Memoization caches will be garbage-collected if they have not been hit after" gcAfterNotHitDuration=30s
11:23:55 workflow-controller.1 | time="2023-05-02T11:23:55.369Z" level=info msg="not enabling pprof debug endpoints"
11:23:55 workflow-controller.1 | time="2023-05-02T11:23:55.441Z" level=info msg="Get configmaps 404"
11:23:55 workflow-controller.1 | time="2023-05-02T11:23:55.442Z" level=fatal msg="Failed to register watch for controller config map: configmaps \"workflow-controller-configmap\" not found"
11:23:55 system                | workflow-controller.1 stopped (rc=1)
11:23:55 system                | sending SIGTERM to argo-server.1 (pid 2626)
11:23:55 system                | sending SIGTERM to ui.1 (pid 2625)
11:23:55 system                | argo-server.1 stopped (rc=-15)
11:23:55 system                | ui.1 stopped (rc=-15)
(devenv) JPs-MacBook-Pro:argo-workflows jpzivalich$
```
I think it might be having a hard time installing things on my docker-desktop cluster
@isubasinghe - could you point me in the right direction?

File: README.md [hunk 1]
```diff
@@ -45,6 +45,19 @@ Argo is a [Cloud Native Computing Foundation (CNCF)](https://cncf.io/) hosted pr
...
 ![Screenshot](docs/assets/screenshot.png)
 
+## Try Argo using Nix
+Nix is a package manager / build tool which focusses on reproducible build environments.
+Argo Workflows has some basic support for Nix which is enough to get Argo Workflows up and running with minimal effort. 
+Here are the steps to follow: 
+  1. modify the hosts file according to [this](https://argoproj.github.io/argo-workflows/running-locally/), don't worry about the other instructions. 
+  2. setup a k8s cluster, k3d is the recommended solution here. 
+  3. Install [Nix](https://nixos.org/download.html) 
+  4. Run "nix develop" (you will have to use the experimental feature flags on a fresh install, the features you need are "nix-command" and "flakes").
```
Comment (JPZ13) [inline README.md]: nit - it'd be better to type the full command here: `nix develop --extra-experimental-features nix-command --extra-experimental-features flakes`

**Comment by JPZ13** [comment]
On our call yesterday @rohankmr414 had a question about adding the dependency files. Rohan, could you ask it here?

File: README.md [hunk 2]
```diff
@@ -45,6 +45,19 @@ Argo is a [Cloud Native Computing Foundation (CNCF)](https://cncf.io/) hosted pr
 
 ![Screenshot](docs/assets/screenshot.png)
 
+## Try Argo using Nix
+Nix is a package manager / build tool which focusses on reproducible build environments.
+Argo Workflows has some basic support for Nix which is enough to get Argo Workflows up and running with minimal effort. 
+Here are the steps to follow:
```
Comment (isubasinghe) [inline README.md:51]: add step to 
"mkdir ./devenv/" as the nix shell doesn't have the permissions to create it.

**Comment by isubasinghe** [comment]
I should have probably mentioned this before, but this does assume kubernetes related stuff has been setup.

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+                    pythonEnv
+                    clang-tools
+                    protobuf
+                    myyarn
+                  ];
+                  enterShell = ''
+                    ./hack/port-forward.sh;
+                    ./hack/free-port 9090;
+                    ./hack/free-port 2746;
+                    ./hack/free-port 8080;
```
Comment (isubasinghe) [inline flake.nix]: you forgot the .sh extension here

Comment (JPZ13) [reply flake.nix]: Are you writing these comments to yourself? 😂

**Comment by JPZ13** [comment]
@isubasinghe and I hopped on a call today and were able to get my issue sorted. All I had to do was restart the terminal after installing nix rather than just opening up a new tab on iterm 🤷‍♂️

File: README.md [hunk 4]
```diff
@@ -45,6 +45,19 @@ Argo is a [Cloud Native Computing Foundation (CNCF)](https://cncf.io/) hosted pr
...
+Here are the steps to follow: 
+  1. modify the hosts file according to [this](https://argoproj.github.io/argo-workflows/running-locally/), don't worry about the other instructions. 
+  2. setup a k8s cluster, k3d is the recommended solution here. 
+  3. Install [Nix](https://nixos.org/download.html) 
+  4. Run "nix develop" (you will have to use the experimental feature flags on a fresh install, the features you need are "nix-command" and "flakes").
+  5. Run "devenv up". 
+  6. That is all you need. 
+
+As I said, this is still fairly barebones at the moment, any feature in the Makefile not mentioned here is excluded at the moment. 
+In practice this means only a "make start UI=true" equivalent is supported at the moment, there are no LDFlags set in the build, so the UI will show something similar to "0.0.0-unknown"
```
Comment (Joibel) [inline README.md:60]: I got to `make start UI=true`, and it fails with `compile: version "go1.20.3" does not match go tool version "go1.19.8"` whilst compiling `dist/workflow-controller`.
I'm on real linux, (arch) and have go installed manually
`PATH="/opt/go/1.20.3/bin:/usr/local/bin:/usr/bin:/home/alan/.krew/bin:/home/alan/.local/bin:/usr/local/sbin:/var/lib/flatpak/exports/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/alan/.krew/bin:/home/alan/.local/bin"` before I enter nix, and `GOROOT="/opt/go/1.20.3"` both in and out. I guess GOROOT is causing this.

Comment (Joibel) [reply README.md:60]: Unsetting GOROOT does the job, so nix should control that environment variable.

File: flake.nix (hunk 3 above)
Comment (isubasinghe) [reply flake.nix]: Lol yes, I wanted to remind myself to fix this.

File: README.md (hunk 4 above)
Comment (isubasinghe) [reply README.md:60]: Thanks @Joibel I shall do that now.

**Review by JPZ13 — APPROVED** [review]
Looks good! 🚀 

I had to delete my `.devenv` folder after pulling the new commits, but after deleting and refreshing, it worked great

**Pull request merged**